File: java/src/main/java/benchmarks/Base64Encode.java
package benchmarks;

import java.util.Base64;
import java.nio.charset.StandardCharsets;

public class Base64Encode extends Benchmark {
    private int n;
    private byte[] bytes;
    private String str2;
    private long resultVal;

    public Base64Encode() {
        n = (int) configVal("size");
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "Base64Encode";
    }

    @Override
    public void prepare() {
        String str = "a".repeat(n);
        bytes = str.getBytes();
        str2 = Base64.getEncoder().encodeToString(bytes);
    }

    @Override
    public void run(int iterationId) {
        str2 = Base64.getEncoder().encodeToString(bytes);
        resultVal = (resultVal + str2.length()) & 0xFFFFFFFFL;
    }

    @Override
    public long checksum() {
        String str = new String(bytes, 0, 5, StandardCharsets.UTF_8);
        String prefix = str.length() > 4 ? str.substring(0, 4) + "..." : str;
        String prefix2 = str2.length() > 4 ? str2.substring(0, 4) + "..." : str2;
        String message = "encode " + prefix + " to " + prefix2 + ": " + resultVal;
        return Helper.checksum(message);
    }
}

File: nim/src/benchmarks/base64encode.nim
import std/[base64, strutils]
import ../benchmark
import ../helper

type
  Base64Encode* = ref object of Benchmark
    str: string
    str2: string
    resultVal: uint32

proc newBase64Encode(): Benchmark =
  Base64Encode()

method name(self: Base64Encode): string = "Base64Encode"

method prepare(self: Base64Encode) =
  let n = self.config_val("size")
  self.str = repeat('a', n.int)

  self.str2 = encode(self.str)
  self.resultVal = 0

method run(self: Base64Encode, iteration_id: int) =
  self.str2 = encode(self.str)
  self.resultVal = self.resultVal + uint32(self.str2.len)

method checksum(self: Base64Encode): uint32 =
  var ss = "encode "
  if self.str.len > 4:
    ss.add(self.str[0..3] & "...")
  else:
    ss.add(self.str)

  ss.add(" to ")

  if self.str2.len > 4:
    ss.add(self.str2[0..3] & "...")
  else:
    ss.add(self.str2)

  ss.add(": " & $self.resultVal)
  checksum(ss)

registerBenchmark("Base64Encode", newBase64Encode)